volatility3/framework/plugins/windows/debugregisters.py

```diff
@@ -0,0 +1,219 @@
+            return None
+
+        return owner_proc, dr7, dr0, dr1, dr2, dr3
+
+    def _get_vads(
+        self,
+        vads_cache: Dict[int, List[Tuple[int, int, str]]],
+        owner_proc: interfaces.objects.ObjectInterface,
+    ) -> Optional[List[Tuple[int, int, str]]]:
+        if owner_proc.vol.offset in vads_cache:
+            vads = vads_cache[owner_proc.vol.offset]
+        else:
+            vads = pe_symbols.PESymbols.get_proc_vads_with_file_paths(owner_proc)
+            vads_cache[owner_proc.vol.offset] = vads
+
+        # smear or terminated process
+        if len(vads) == 0:
+            return None
+
+        return vads
+
+    def _generator(
+        self,
+    ) -> Generator[
+        Tuple[
+            int,
+            Tuple[
+                str,
+                int,
+                int,
+                int,
+                int,
+                format_hints.Hex,
+                str,
+                str,
+                format_hints.Hex,
+                str,
+                str,
+                format_hints.Hex,
+                str,
+                str,
+                format_hints.Hex,
+                str,
+                str,
+            ],
+        ],
+        None,
+        None,
+    ]:
+        kernel = self.context.modules[self.config["kernel"]]
+
+        vads_cache: Dict[int, List[Tuple[int, int, str]]] = {}
+
+        proc_modules = None
+
+        procs = pslist.PsList.list_processes(
+            context=self.context,
+            layer_name=kernel.layer_name,
+            symbol_table=kernel.symbol_table_name,
+        )
+
+        for proc in procs:
+            for thread in threads.Threads.list_threads(kernel, proc):
+                debug_info = self._get_debug_info(thread)
+                if not debug_info:
+                    continue
+
+                owner_proc, dr7, dr0, dr1, dr2, dr3 = debug_info
+
+                vads = self._get_vads(vads_cache, owner_proc)
+                if not vads:
+                    continue
+
+                # this lookup takes a while, so only perform if we need to
+                if not proc_modules:
+                    proc_modules = pe_symbols.PESymbols.get_process_modules(
+                        self.context, kernel.layer_name, kernel.symbol_table_name, None
+                    )
+                    path_and_symbol = partial(
+                        pe_symbols.PESymbols.path_and_symbol_for_address,
+                        self.context,
+                        self.config_path,
+                        proc_modules,
+                    )
+
+                file0, sym0 = path_and_symbol(vads, dr0)
+                file1, sym1 = path_and_symbol(vads, dr1)
+                file2, sym2 = path_and_symbol(vads, dr2)
+                file3, sym3 = path_and_symbol(vads, dr3)
+
+                # if none map to an actual file VAD then bail
+                if not (
+                    isinstance(file0, str)
```

Comment:
Is there a better way to check for it being a str or one of the `renders.Not*` types? This pattern is many places in plugins already though.

Reply:
It is used in a few places, but using type information to choose code-paths seems like a bad idea (particularly if we ever want to reimplement this in a typed language, for instance) so I'm trying to keep the number of usages down..  Given it's a function we control, we could easily return a boolean to say whether it succeeded (or just return `None` rather than a partial tuple).

Reply:
The file#, sym# variables still need to get rendered even if they aren't found though in the cases where at least 1 did map. So if `path_and_symbol_for_address` returned None in the cases of failure, then we would need additional checks after the calls to say:

```
if file0 is None:
   file0 = renderers.NotAvailable()
   
if file1...

if file2...

if file3...
```

It seems alot cleaner to let a str or a render.* type return then check the instance unless/until some other method is made to check if a value is a real string or not.

Reply:
Pretty sure when filling the tuple you could go `variable or NotAvailable` and if the variable is None, you end up with the AbsentValue?

Reply:
No, the point is just before it gets yielded, it gets yielded with a `variable or NotAvailable` that changes the `None`s into `NotAvailable`s.  `None`s directly into the TreeGrid will still break.  So for instance, change line 176 from `file0,` to `file0 or renderers.NotAvailable(),`.  That way, all of python sees it as a `None`, but the `TreeGrid` gets to see it as a `NotAvailable`...

Reply:
Here's an example, although *annoyingly* it was done in a subfunction rather than just before use in the yield...  5:S

https://github.com/volatilityfoundation/volatility3/blob/d56cd83510e64f9f875ff1dad11d8c8cbf5995f5/volatility3/framework/plugins/linux/lsof.py#L56